task_manager/task_manager.cc

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "task_manager.hh"
+#include "db/config.hh"
+
+seastar::gate& task_manager::async_gate() {
+    return _gate;
+}
+
+abort_source& task_manager::abort_source() {
+    return _as;
+}
+
+std::chrono::seconds task_manager::task_ttl() const {
+    return std::chrono::seconds(_cfg.task_ttl_in_seconds());
+}
+
+task_manager::tasks_per_module& task_manager::get_tasks_per_module() {
+    return _tasks_per_module;
+}
+
+task_manager::task_map& task_manager::get_all_tasks() {
+    return _all_tasks;
+}
+
+task_manager::task_ptr task_manager::get_task_this_shard(task_id id) const {
+    auto it = _all_tasks.find(id);
+    if (it != _all_tasks.end()) {
+        return it->second;
+    }
+    return nullptr;
+}
+
+std::vector<std::string> task_manager::get_modules() const {
+    return {"repair", "compaction"}; // TODO Deexie change to _modules
+}
+
+void task_manager::register_task(task_ptr task) {
+    auto id = task->get_status().id;
+    _tasks_per_module[task->get_module()].insert({id, task});
+    _all_tasks[id] = task;
+}
+
+void task_manager::unregister_task(task_id id) {
+    if (_all_tasks.contains(id)) {
+        std::string module = _all_tasks[id]->get_module();
+        _tasks_per_module[module].erase(id);
+        _all_tasks.erase(id);
+    }
+}
+
+task_manager::task_ptr task_manager::retrieve_task_this_shard(task_id id) {
+    task_ptr task = get_task_this_shard(id);
+    if (task == nullptr) {
+        return nullptr;
+    }
+    auto state = task->get_status().state;
+    if (state == task_manager::task_state::done || state == task_manager::task_state::failed) {
+        unregister_task(id);
+    }
+    return task;
+}
+
+future<task_manager::task_ptr> task_manager::wait_task_this_shard(task_id id) {
+    task_ptr task = get_task_this_shard(id);
+    if (task == nullptr) {
+        return make_ready_future<task_ptr>(nullptr);
+    }
+    return task->done().then([this, task, id] {
+        unregister_task(id);
+        return make_ready_future<task_ptr>(task);
+    });
+}
+
+future<> task_manager::stop() {
+    return _gate.close(); // TODO Deexie split stopping ionto modules (currently all modules are stopped together) 
+}
+
+future<unsigned> task_manager::get_task_shard(task_id id) const {
+    return container().map_reduce0([id] (const task_manager& tm) {
+        auto task = tm.get_task_this_shard(id);
+        if (task == nullptr) {
+            return smp::count;
+        }
+        return task->get_status().shard;
+    }, smp::count, [] (unsigned l, unsigned r) -> unsigned {
+        return std::min(l, r);
+    });
+}
```

Comment:
What's the use case for this function?
Ideally it would be better introduced in the patch series right before or along with the use case.